// @bitdev/node.generators.node-templates@1.0.9/dist/platform/files/node-app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeAppFile = void 0;
const nodeAppFile = (context) => {
    const { name, namePascalCase } = context;
    return {
        relativePath: `${name}.bit-app.tsx`,
        content: `import { Platform } from '@bitdev/platforms.platform';

const UserServer = import.meta.resolve('@bitdev/node.examples.user-server');
const DiscussionServer = import.meta.resolve('@bitdev/platforms.examples.discussion-server');
const AcmeWeb = import.meta.resolve('@bitdev/platforms.examples.acme-web');
const PlatformGateway = import.meta.resolve('@bitdev/symphony.backends.gateway-server');

export const ${namePascalCase} = Platform.from({
  name: '${name}',

  frontends: {
    main: AcmeWeb,
    mainPortRange: [3000, 3100]
  },

  backends: {
    // use the default gateway component. supports proxy of graphql and restful requests.
    main: PlatformGateway,
    services: [
      // composes example micro-services. replace with your own.
      UserServer,
      DiscussionServer
    ]
  },
});

export default ${namePascalCase};
`,
    };
};
exports.nodeAppFile = nodeAppFile;
