// @bitdesign/sparks.content.logo@0.0.4/logo.tsx
import type { ReactNode } from 'react';
import classNames from 'classnames';
import { Flex } from '@bitdesign/sparks.layout.flex';
import { Link } from '@bitdesign/sparks.navigation.link';
import styles from './logo.module.scss';

export type LogoProps = {
  /**
   * children to render below the logo.
   */
  children?: ReactNode;

  /**
   * name of the logo.
   */
  displayText?: boolean;

  /**
   * custom size of the logo.
   */
  imageSize?: number;

  /**
   * name of the logo.
   */
  name?: string;

  /**
   * slogan for the logo.
   */
  slogan?: string;

  /**
   * source for the logo.
   */
  src?: string;

  /**
   * set the logo as a link.
   */
  href?: string;

  /**
   * logo text class.
   */
  logoTextClass?: string;

  /**
   * class name to use.
   */
  className?: string;

  /**
   * alt text for the logo.
   */
  alt?: string;
};

export function Logo({ children, alt, logoTextClass, href, src, displayText = true, name, className, slogan, imageSize = 32 }: LogoProps) {
  const flex = (
    <Flex className={classNames(className, styles.logo)}>
      {src && <div>
        <img height={imageSize} src={src} alt={alt || name} />
      </div>}
      {displayText && <Flex className={classNames(styles.logoText, logoTextClass)}>
        <b>{name}</b>
        {slogan && <span className={styles.slogan}>{slogan}</span>}
      </Flex>}
      {children}
    </Flex>
  );

  if (href) return <Link className={styles.link} href={href}>{flex}</Link>;
  
  return flex;
}
