// @bitdev/node.generators.node-starters@0.0.6/dist/starters/node-workspace/node-workspace-starter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeWorkspaceStarter = void 0;
const component_id_1 = require("@teambit/component-id");
const workspace_common_1 = require("../../workspace-common");
const constants_1 = require("../../workspace-common/constants");
class NodeWorkspaceStarter {
    constructor(name = 'platform', description = 'Platform workspace with demo components', hidden = false) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
    }
    generateFiles(context) {
        return __awaiter(this, void 0, void 0, function* () {
            const appId = this.getAppId(context);
            return (0, workspace_common_1.generateFiles)(context, {
                [appId.toStringWithoutVersion()]: {}
            }, ['bitdev.react/react-env']);
        });
    }
    getAppId(context) {
        const projectName = context.name;
        const defaultScope = context.defaultScope || constants_1.DEFAULT_SCOPE;
        return component_id_1.ComponentID.fromString(`${defaultScope}/${projectName}-platform`);
    }
    fork(context) {
        const projectName = context.name;
        const appId = this.getAppId(context);
        const defaultScope = context.defaultScope || constants_1.DEFAULT_SCOPE;
        const [owner] = defaultScope.split('.');
        return [
            {
                id: 'acme.acme/layout/header',
                targetName: `layout/header`
            },
            {
                id: 'acme.acme/acme-web',
                targetName: `${projectName}-web`,
            },
            {
                id: 'acme.acme/acme-platform',
                targetName: appId.fullName
            },
            {
                id: 'acme.discussions/discussion-feed',
                targetName: `discussion-feed`,
                targetScope: `${owner}.discussions`
            },
            {
                id: 'acme.discussions/discussions-server',
                targetName: 'discussions-server',
                targetScope: `${owner}.discussions`
            },
            {
                id: 'acme.discussions/entities/discussion',
                targetName: 'entities/discussion',
                targetScope: `${owner}.discussions`
            },
            {
                id: 'acme.discussions/hooks/use-discussion',
                targetName: 'hooks/use-discussion',
                targetScope: `${owner}.discussions`
            }
        ];
    }
    static from(options = {}) {
        return () => new NodeWorkspaceStarter(options.name, options.description, options.hidden);
    }
}
exports.NodeWorkspaceStarter = NodeWorkspaceStarter;
