// @bitdev/node.generators.node-templates@1.0.9/dist/entity-template/entity-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityTemplate = void 0;
const index_file_1 = require("./files/index-file");
const component_1 = require("./files/component");
const test_1 = require("./files/test");
const docs_1 = require("./files/docs");
class EntityTemplate {
    constructor(name = 'entity', description = 'create entities to reuse data structures efficiently', hidden = false, env) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, component_1.componentFile)(context),
            (0, test_1.testFile)(context),
            (0, docs_1.docsFile)(context),
        ];
    }
    static from(options = {}) {
        return () => new EntityTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.EntityTemplate = EntityTemplate;
