// @bitdesign/sparks.foundations.color-palette@0.0.1/dist/color-palette.composition.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import React from 'react';
import { ColorPalette } from './color-palette.js';
const ColorPaletteGrid = ({ colorPalette }) => {
    return (_jsx("div", { style: { display: 'flex', flexDirection: 'column', gap: 15, padding: 15 }, children: colorPalette.colors.map((color, key) => {
            const tones = color.listTones();
            return (_jsxs("div", { style: { display: 'flex', gap: 7.5 }, children: [_jsx("span", { children: color.name }), tones.map((tone, toneKey) => _jsx("div", { style: { backgroundColor: tone.value, width: 50, height: 50, borderRadius: 8 }, children: tone.name }, toneKey))] }, key));
        }) }));
};
export const DefaultColorPalette = () => {
    const colorPalette = ColorPalette.create();
    return _jsx(ColorPaletteGrid, { colorPalette: colorPalette });
};
export const GreenPrimaryColor = () => {
    const colorPalette = ColorPalette.create({
        primary: {
            origin: '#1E884A'
        }
    });
    return _jsx(ColorPaletteGrid, { colorPalette: colorPalette });
};
export const PurplePrimaryColor = () => {
    const colorPalette = ColorPalette.create({
        primary: {
            origin: '#6d3e90'
        }
    });
    return _jsx(ColorPaletteGrid, { colorPalette: colorPalette });
};
export const YellowPrimaryColor = () => {
    const colorPalette = ColorPalette.create({
        primary: {
            origin: '#FFF961'
        }
    });
    return _jsx(ColorPaletteGrid, { colorPalette: colorPalette });
};
export const LightPurplePrimaryColor = () => {
    const colorPalette = ColorPalette.create({
        primary: {
            origin: '#D185EC'
        },
        error: {
            origin: '#F90069'
        }
    });
    return _jsx(ColorPaletteGrid, { colorPalette: colorPalette });
};
