// @bitdev/node.generators.node-templates@1.0.9/dist/node-env-template/files/docs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.envDocsFile = void 0;
function envDocsFile(context) {
    return {
        relativePath: `${context.name}.docs.mdx`,
        content: `---
description: 'A NodeJS development environment.'
labels: ['env', 'node', 'dev', 'mocha', 'typescript']
---

A reusable development environment for NodeJS.  
Use the NodeJS environment to create new NodeJS modules, entities and apps.

Features:
- ▶️ TypeScript
- 🧪 Mocha.
- 🧑‍💻 Generator templates include [NodeJS Module](/), [Entity](/), [Express App](/), [Platform app](/) and a [native Bit app](/).
- 🚀 Workspace starter
- 🦋 ESLint
- ⚒️ A NodeJS optimized build pipeline.
- 👓 React-based Preview

## Use in existing workspaces

Add the development env to your generator configuration on \`workspace.jsonc\`.  

\`\`\`json
{
  "teambit.generator/generator": {
    "envs": [
      "${context.componentId.toStringWithoutVersion()}"
    ]
  }
}
\`\`\`

You can see the generator templates in your workspace using \`bit templates\`.

## Use on an existing component

Use this dev environment on your existing components, by running the following command: 

\`\`\`
bit env set my-component ${context.componentId.toStringWithoutVersion()}
\`\`\`


## Create a NodeJS workspace

You can create a new NodeJS workspace made from this env using the following command:

\`\`\`
bit new node my-workspace --aspect ${context.componentId.toStringWithoutVersion()}
\`\`\`

## Compose your own env

You can base on this env, and customize it to your needs. To create your env run this command:
\`\`\`
bit create node-env my-new-env
\`\`\`

Make sure your new environment class extends this env class and overrides the needed class methods.
`,
    };
}
exports.envDocsFile = envDocsFile;
