// @bitdesign/sparks.layout.header@0.0.5/header.tsx
import type { ComponentType, ReactNode } from 'react';
import { Flex, type FlexProps } from '@bitdesign/sparks.layout.flex';
import classNames from 'classnames';
import styles from './header.module.scss';

export type HeaderProps = {
  /**
   * set the children for the header. 
   * used in the center of the layout.
   */
  children?: ReactNode;

  /**
   * logo to use.
   */
  logo?: ReactNode;

  /**
   * badges to include at the left of the header. 
   */
  actions?: ComponentType[];

  /**
   * class name for the header.
   */
  className?: string;

  /**
   * class name for the content container.
   */
  contentClassName?: string;

  /**
   * class name to use for logo container.
   */
  logoClassName?: string;

  /**
   * class name to use for badges.
   */
  actionsClassName?: string;
} & FlexProps;

export function Header({ children, actions, logo, contentClassName, logoClassName, actionsClassName, className, ...rest }: HeaderProps) {
  return (
    <Flex {...rest} justifyContent='space-between' className={classNames(styles.header, className)}>
      <Flex className={classNames(styles.left)}>
        {logo && <div className={styles.logo}>{logo}</div>}
        {children}
      </Flex>
      {actions && <Flex className={classNames(actionsClassName)}>
        {actions.map((badge, key) => {
          const BadgeComponent = badge;
          return <BadgeComponent key={key} />;
        })}
      </Flex>}
    </Flex>
  );
}
