// @bitdev/react.generators.react-starters@0.0.8/dist/starters/react-workspace/react-workspace-starter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactWorkspaceStarter = void 0;
const component_id_1 = require("@teambit/component-id");
const workspace_common_1 = require("../../workspace-common");
const constants_1 = require("../../workspace-common/constants");
class ReactWorkspaceStarter {
    constructor(name = 'react', description = 'React workspace with demo components', hidden = false) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
    }
    generateFiles(context) {
        return __awaiter(this, void 0, void 0, function* () {
            const appId = this.getAppId(context);
            return (0, workspace_common_1.generateFiles)(context, {
                [appId.toStringWithoutVersion()]: {}
            }, ['bitdev.node/node-env']);
        });
    }
    getAppId(context) {
        const projectName = context.name;
        const defaultScope = context.defaultScope || constants_1.DEFAULT_SCOPE;
        // const [owner] = defaultScope.split('.');
        return component_id_1.ComponentID.fromString(`${defaultScope}/${projectName}`);
    }
    fork(context) {
        const projectName = context.name;
        const appId = this.getAppId(context);
        return [
            {
                id: 'acme.corporate/corporate-website',
                targetName: `${projectName}-web`
            },
            {
                id: 'acme.corporate/corporate-service',
                targetName: `${projectName}-service`,
            },
            {
                id: 'acme.corporate/acme-corporate',
                targetName: appId.fullName
            },
            {
                id: 'acme.corporate/ui/announcements',
                targetName: `ui/announcements`
            },
            {
                id: 'acme.corporate/hooks/use-announcements',
                targetName: 'hooks/use-announcements',
            },
            {
                id: 'acme.corporate/entities/announcement',
                targetName: 'entities/announcement'
            },
        ];
    }
    static from(options = {}) {
        return () => new ReactWorkspaceStarter(options.name, options.description, options.hidden);
    }
}
exports.ReactWorkspaceStarter = ReactWorkspaceStarter;
