// @bitdev/node.node-server@0.0.9/node-server-options.ts
import type { BuildFn, DeployFn } from "@teambit/application";
import { BuildOptions } from "esbuild";

export type NodeServerOptions = {
  /**
   * name of the node app.
   */
  name: string;

  /**
   * path to the app main file.
   */
  mainPath: Promise<string>|string;

  /**
   * build task.
   */
  build?: BuildFn;

  /**
   * deploy task.
   */
  deploy?: DeployFn;

  /**
   * path or name of a binary in your path to nodejs.
   * defaults to `node`
   */
  nodeBin?: string;

  /**
   * name of the artifact. defaults to 'app-bundle'
   */
  artifactName?: string;

  /**
   * filename used for the bundle.
   * defaults to {name}.js
   */
  bundleFileName?: string;

  /**
   * a single binary execution file.
   */
  binary?: boolean;

  /**
   * options for esbuild to use during build.
   */
  esbuildOptions?: BuildOptions,

  /**
   * port range to use.
   */
  portRange?: [number, number],

  /**
   * determine whether to show server
   * name on the outputs.
   */
  showNameInOutputs?: boolean,

  /**
   * dist dir to use.
   * defaults to `dist`
   */
  distDir?: string,
};
