// @bitdev/node.generators.node-templates@1.0.9/dist/node-app/node-app-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeAppTemplate = void 0;
const index_file_1 = require("./files/index-file");
const component_1 = require("./files/component");
const test_1 = require("./files/test");
const docs_1 = require("./files/docs");
const bit_app_1 = require("./files/bit-app");
class NodeAppTemplate {
    constructor(name = 'node-app', description = 'create a basic nodejs app.', hidden = false, env, isApp = true, installMissingDependencies = true) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
        this.isApp = isApp;
        this.installMissingDependencies = installMissingDependencies;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, component_1.componentFile)(context),
            (0, test_1.testFile)(context),
            (0, docs_1.docsFile)(context),
            (0, bit_app_1.bitAppFile)(context)
        ];
    }
    static from(options = {}) {
        return () => new NodeAppTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.NodeAppTemplate = NodeAppTemplate;
