// @bitdesign/sparks.sparks-theme@0.0.1/create-tokens.ts
import { Spacing } from "@bitdesign/sparks.foundations.spacing";
import { ColorPalette, sparksColorPalette } from "@bitdesign/sparks.foundations.color-palette";
import { LoadTokensContext } from "./tokens-context.js";

export function createTokens(context?: LoadTokensContext) {
  const palette = context?.colorPalette || ColorPalette.create(sparksColorPalette);
  const spacing = context?.spacing || Spacing.create();

  const tokens = {
    // spacing: spacing.createTokens(),

    // palette: palette.toObject(),

    /**
     * background color. use for primary surfaces across your design.
     */
    backgroundColor: palette.neutral[0],

    /**
     * fill colors.
     */
    fill: {
      /**
       * default fill color.
       */
      default: palette.neutral[100],

      /**
       * primary fill colors.
       */
      primary: {
        /**
         * default primary color.
         */
        default: palette.primary[600],

        /**
         * primary color for hover surfaces.
         */
        hover: palette.primary.dark,
      },

      /**
       * secondary fill colors.
       */
      secondary: {
        /**
         * default color for secondary fills.
         */
        default: palette.neutral[200],

        /**
         * hover color for secondary fills.
         */
        hover: palette.neutral[300]
      },
    },

    /**
     * content tokens.
     */
    content: {
      default: palette.neutral[1000],
      inverse: palette.neutral[0],
      primary: palette.primary.default,

      /**
       * content size.
       */
      size: '16px',
    },
    
    border: {
      default: palette.neutral[200],
      focus: palette.primary.light,
      radius: spacing.default.px
    },
  
    /**
     * fill selected color
     */
    fillSelected: palette.primary[200],
  
    spacing: {
      default: spacing.default.px,
      half: spacing.half.px,
      double: spacing.double.px,
      triple: spacing.triple.px,
      x4: spacing.get(400).px,
      x6: spacing.get(600).px,
    },

    /**
     * color for secondary surfaces on the screen such 
     * as cards or content outlius.
     */
    // fillDefault: palette.neutral[100],

    /**
     * transition ease in.
     */
    transitionEaseIn: '0.3s ease-in',

    /**
     * transition ease out.
     */
    transitionEaseOut: '0.3s ease-out',
  };

  return tokens;
}
