// @bitdev/node.generators.node-templates@1.0.9/dist/node-env-template/node-env-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeEnvTemplate = void 0;
const index_file_1 = require("./files/index-file");
const node_env_bit_env_1 = require("./files/node-env-bit-env");
const env_jsonc_1 = require("./files/env-jsonc");
const eslintrc_1 = require("./files/config/eslintrc");
const tsconfig_json_1 = require("./files/config/tsconfig.json");
const jest_config_1 = require("./files/config/jest.config");
const prettier_config_1 = require("./files/config/prettier.config");
const webpack_transformer_1 = require("./files/config/webpack-transformer");
const asset_1 = require("./files/types/asset");
const style_1 = require("./files/types/style");
const docs_1 = require("./files/docs");
class NodeEnvTemplate {
    constructor(name = 'node-env', description = 'create a customized nodejs dev environment', hidden = false) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, node_env_bit_env_1.NodeEnv)(context),
            (0, env_jsonc_1.envJson)(context),
            (0, eslintrc_1.eslintConfigFile)(),
            (0, tsconfig_json_1.tsConfigFile)(),
            (0, jest_config_1.jestConfigFile)(),
            (0, prettier_config_1.prettierConfigFile)(),
            (0, webpack_transformer_1.webpackTransformerFile)(),
            (0, asset_1.assetsFile)(),
            (0, style_1.styleFile)(),
            (0, docs_1.envDocsFile)(context)
        ];
    }
    config() {
        return {
            ['teambit.envs/envs']: {
                env: 'teambit.envs/env',
            },
        };
    }
    static from(options = {}) {
        return () => new NodeEnvTemplate(options.name, options.description, options.hidden);
    }
}
exports.NodeEnvTemplate = NodeEnvTemplate;
