// @bitdev/node.generators.node-templates@1.0.9/dist/graphql-template/graphql-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLServerTemplate = void 0;
const index_file_1 = require("./files/index-file");
const node_app_1 = require("./files/node-app");
const app_root_1 = require("./files/app-root");
const app_1 = require("./files/app");
const docs_1 = require("./files/docs");
const app_test_1 = require("./files/app-test");
const graphql_file_1 = require("./files/graphql-file");
class GraphQLServerTemplate {
    constructor(name = 'graphql-server', description = 'create a basic apollo graphql server', hidden = false, env, isApp = true, installMissingDependencies = true) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
        this.isApp = isApp;
        this.installMissingDependencies = installMissingDependencies;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, docs_1.docsFile)(context),
            (0, app_test_1.appSpecFile)(context),
            (0, node_app_1.nodeAppFile)(context),
            (0, app_1.expressAppFile)(context),
            (0, graphql_file_1.graphqlFile)(context),
            (0, app_root_1.AppRootFile)(context),
        ];
    }
    static from(options = {}) {
        return () => new GraphQLServerTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.GraphQLServerTemplate = GraphQLServerTemplate;
