// @bitdev/node.generators.node-starters@0.0.6/starters/node-workspace/node-workspace-starter.ts
import { WorkspaceContext, WorkspaceTemplate } from '@teambit/generator';
import { ComponentID } from '@teambit/component-id';
import { generateFiles as generateCommonFiles } from '../../workspace-common';
import { DEFAULT_SCOPE } from '../../workspace-common/constants';

export type ReactComponentTemplateOptions = {
  /**
   * name of the template
   */
  name?: string;

  /**
   * description of the template.
   */
  description?: string;

  /**
   * hide the template from the `bit templates` command.
   */
  hidden?: boolean;
};

export class NodeWorkspaceStarter implements WorkspaceTemplate {
  constructor(
    readonly name = 'platform',
    readonly description = 'Platform workspace with demo components',
    readonly hidden = false
  ) {}

  async generateFiles(context: WorkspaceContext) {
    const appId = this.getAppId(context);
    
    return generateCommonFiles(context, {
      [appId.toStringWithoutVersion()]: {} 
    }, ['bitdev.react/react-env']);
  }

  getAppId(context: WorkspaceContext) {
    const projectName = context.name;
    const defaultScope = context.defaultScope || DEFAULT_SCOPE;

    return ComponentID.fromString(`${defaultScope}/${projectName}-platform`);
  }

  fork(context: WorkspaceContext) {
    const projectName = context.name;
    const appId = this.getAppId(context);
    const defaultScope = context.defaultScope || DEFAULT_SCOPE;
    const [owner] = defaultScope.split('.');

    return [
      {
        id: 'acme.acme/layout/header',
        targetName: `layout/header`
      },
      {
        id: 'acme.acme/acme-web',
        targetName: `${projectName}-web`,
      },
      {
        id: 'acme.acme/acme-platform',
        targetName: appId.fullName
      },
      {
        id: 'acme.discussions/discussion-feed',
        targetName: `discussion-feed`,
        targetScope: `${owner}.discussions`
      },
      {
        id: 'acme.discussions/discussions-server',
        targetName: 'discussions-server',
        targetScope: `${owner}.discussions`
      },
      {
        id: 'acme.discussions/entities/discussion',
        targetName: 'entities/discussion',
        targetScope: `${owner}.discussions`
      },
      {
        id: 'acme.discussions/hooks/use-discussion',
        targetName: 'hooks/use-discussion',
        targetScope: `${owner}.discussions`
      }
    ];
  }

  static from(options: ReactComponentTemplateOptions = {}) {
    return () =>
      new NodeWorkspaceStarter(
        options.name,
        options.description,
        options.hidden
      );
  }
}
