// @bitdev/node.generators.node-templates@1.0.9/dist/platform/platform-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformTemplate = void 0;
const index_file_1 = require("./files/index-file");
const node_app_1 = require("./files/node-app");
class PlatformTemplate {
    constructor(name = 'platform', description = 'compose a platform from services and apps', hidden = false, env, isApp = true, installMissingDependencies = true) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
        this.isApp = isApp;
        this.installMissingDependencies = installMissingDependencies;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, node_app_1.nodeAppFile)(context),
            // expressAppFile(context),
        ];
    }
    static from(options = {}) {
        return () => new PlatformTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.PlatformTemplate = PlatformTemplate;
