// @acme/design.acme-theme@2.0.2/acme-tokens.ts
/**
 * Acme tokens theme.
 * Include all tokens in this object.
 */
export const acmeTokens = () => {
  return {
    /**
     * background color. use for primary surfaces across your design.
     */
    backgroundColor: '#fffff',

    /**
     * color for general purpose text color. expected to mainly apply
     * on `backgroundColor`
     */
    textColor: 'black',

    /**
     * primary brand color. used for primary call to actions.
     * such as buttons, links, etc.
     */
    primaryColor: 'purple',

    /**
     * color to use for borders.
     */
    borderColor: '#ededed',

    /**
     * color for secondary surfaces on the screen such
     * as cards or content outlines.
     */
    surfaceColor: '#f5f5f5',
  };
};

// create a theme type schema to allow new theme to override
// or implement a different theme variation like dark theme.
// in case tokens are dynamically loaded from a json file, please declare this variable an an interface.
export type AcmeThemeSchema = ReturnType<typeof acmeTokens>;
