// @bitdesign/sparks.foundations.color-palette@0.0.1/color-palette.composition.tsx
import React from 'react';
import { ColorPalette } from './color-palette.js';

const ColorPaletteGrid = ({colorPalette}: {colorPalette: ColorPalette}) => {
  return (
    <div style={{ display: 'flex', flexDirection: 'column', gap: 15, padding: 15 }}>
      {colorPalette.colors.map((color, key) => {
      const tones = color.listTones();

      return (
        <div key={key} style={{ display: 'flex', gap: 7.5 }}>
          <span>{color.name}</span>
          {tones.map((tone, toneKey) => <div key={toneKey} style={{ backgroundColor: tone.value, width: 50, height: 50, borderRadius: 8  }}>{tone.name}</div>)}
        </div>
      )
    })}
    </div>
  )
};

export const DefaultColorPalette = () => {
  const colorPalette = ColorPalette.create();

  return <ColorPaletteGrid colorPalette={colorPalette} />;
};

export const GreenPrimaryColor = () => {
  const colorPalette = ColorPalette.create({
    primary: {
      origin: '#1E884A'
    }
  });

  return <ColorPaletteGrid colorPalette={colorPalette} />;
};


export const PurplePrimaryColor = () => {
  const colorPalette = ColorPalette.create({
    primary: {
      origin: '#6d3e90'
    }
  });

  return <ColorPaletteGrid colorPalette={colorPalette} />;
};

export const YellowPrimaryColor = () => {
  const colorPalette = ColorPalette.create({
    primary: {
      origin: '#FFF961'
    }
  });

  return <ColorPaletteGrid colorPalette={colorPalette} />;
};

export const LightPurplePrimaryColor = () => {
  const colorPalette = ColorPalette.create({
    primary: {
      origin: '#D185EC'
    },
    error: {
      origin: '#F90069'
    }
  });

  return <ColorPaletteGrid colorPalette={colorPalette} />;
};
