// @bitdev/harmony.harmony@1.0.13/extension-graph/from-extension.ts
import type { ExtensionManifest } from '../extension/extension-manifest.js';
import { extensionFactory } from '../factory/extension-factory.js';
import ExtensionPotentialCircular from '../exceptions/extension-potential-circular.js';
import { Extension } from '../extension/extension.js';
import { DependencyGraphOptions, Edge } from './extension-graph.js';

function getName(manifest: any) {
  return manifest.id || manifest.name;
}

/**
 * build vertices and edges from the given extension
 */
export function fromExtension(extension: ExtensionManifest, options: DependencyGraphOptions = {}) {
  const vertices: { [id: string]: Extension } = {};
  let edges: { sourceId: string; targetId: string; edge: Edge }[] = [];

  function iterate(root: ExtensionManifest) {
    const id = options.getName ? options.getName(root) : getName(root);
    if (vertices[id]) return;

    const instance = extensionFactory(root);
    const validDeps = instance.dependencies.filter(dep => dep).map(dep => extensionFactory(dep));
    if (instance.dependencies.length > validDeps.length) {
      throw new ExtensionPotentialCircular(instance, validDeps);
    }
    vertices[id] = instance;
    const newEdges = validDeps.map(dep => {
      return {
        sourceId: id,
        targetId: dep.name,
        edge: {
          type: 'dependency'
        }
      };
    });

    edges = edges.concat(newEdges);

    // @ts-ignore
    instance.dependencies.forEach(dep => iterate(dep));
  }

  iterate(extension);

  // eslint-disable-next-line prefer-const
  let vertexArray: { id: string; node: Extension }[] = [];
  // eslint-disable-next-line no-restricted-syntax
  for (const [key, value] of Object.entries(vertices)) {
    vertexArray.push({ id: key, node: value });
  }

  return {
    vertices: vertexArray,
    edges
  };
}

/**
 * build vertices and edges from the given list of extensions
 */
export function fromExtensions(extensions: ExtensionManifest[], options: DependencyGraphOptions = {}) {
  const perExtension = extensions.map(ext => fromExtension(ext, options));

  return perExtension.reduce(
    (acc, subgraph) => {
      acc.edges = acc.edges.concat(subgraph.edges);
      acc.vertices = acc.vertices.concat(subgraph.vertices);

      return acc;
    },
    { vertices: [], edges: [] }
  );
}
