// @backend/kubernetes.examples.acme-web@0.1.0/dist/acme-web.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Outlet, Link, useLocation, useNavigate, } from 'react-router-dom';
import { NavigationProvider, } from '@bitdesign/sparks.navigation.link';
import { Header } from '@acme/acme.layout.header';
import { DiscussionFeed } from '@acme/discussions.discussion-feed';
import { AcmeTheme } from '@acme/design.acme-theme';
import { Card } from '@acme/design.content.card';
const routerAdapter = {
    Link: ({ href, ...props }) => _jsx(Link, { to: href, ...props }),
    useLocation,
    useNavigate,
};
function Layout() {
    return (_jsx(AcmeTheme, { children: _jsxs(NavigationProvider, { adapter: routerAdapter, children: [_jsx(Header, {}), _jsx(Outlet, {})] }) }));
}
export const routes = [
    {
        path: '/',
        element: _jsx(Layout, {}),
        children: [
            {
                index: true,
                element: _jsx(Card, { children: "Welcome to Acme." }),
            },
            {
                path: '/discussions',
                element: _jsx(DiscussionFeed, {}),
            },
        ],
    },
];
